sdks/python/apache_beam/runners/dataflow/internal/apiclient.py

```diff
@@ -588,11 +588,25 @@ def _stage_resources(self, pipeline, options):
               'Found duplicated artifact: %s (%s)',
               type_payload.path,
               type_payload.sha256)
+          staged_name = hashs[type_payload.sha256]
           dep.role_payload = beam_runner_api_pb2.ArtifactStagingToRolePayload(
-              staged_name=hashs[type_payload.sha256]).SerializeToString()
+              staged_name=staged_name).SerializeToString()
         else:
-          resources.append((type_payload.path, role_payload.staged_name))
-          hashs[type_payload.sha256] = role_payload.staged_name
+          staged_name = role_payload.staged_name
+          resources.append((type_payload.path, staged_name))
+          hashs[type_payload.sha256] = staged_name
+
+        if google_cloud_options.staging_location.startswith('gs://'):
+          dep.type_urn = common_urns.artifact_types.URL.urn
+          dep.type_payload = beam_runner_api_pb2.ArtifactUrlPayload(
+              url=FileSystems.join(
+                  google_cloud_options.staging_location,
+                  staged_name)).SerializeToString()
+        else:
+          dep.type_payload = beam_runner_api_pb2.ArtifactFilePayload(
```

Comment:
Can you clarify when a Dataflow pipeline will use "ArtifactFilePayload" instead of "ArtifactUrlPayload" ?

Reply:
AFAIK, ArtifactFilePayload is used for the artifact on the local disk meaning that network connection is not necessary to access the artifact. ArtifactUrlPayload is used when we need to download the artifact in order to access it.

Reply:
Also, keeping the hash around for validation (as we can do with file) is good too.

Reply:
@robertwb We have separate types for PyPI and Maven central artifacts. Do you mean we should use URL artifacts instead?

Reply:
> Also, keeping the hash around for validation (as we can do with file) is good too.

Done. optional sha256 field added to URL payload.

Reply:
Not resolved. I don't think we should be special-casing GCS here and turning it into a URL artifact (and then, presumably on the other end, recognising GCS "urls" to download them).

(With regard to Maven/PyPI, etc. those wouldn't have URLs but dependency specs with version ranges that could be merged.)